packages/opentelemetry-api/src/api/diag.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  API_BACKWARDS_COMPATIBILITY_VERSION,
+  GLOBAL_DIAG_LOGGER_API_KEY,
+  makeGetter,
+  _global,
+} from './global-utils';
+
+/** Internal simple Noop Diag API that returns a noop logger and does not allow any changes */
+function noopDiagApi(): DiagAPI {
+  const noopApi = noopDiagLogger() as DiagAPI;
+
+  noopApi.getLogger = () => noopApi;
+  noopApi.setLogger = noopApi.getLogger;
+  noopApi.setLogLevel = () => {};
+
+  return noopApi;
+}
+
+/**
+ * Singleton object which represents the entry point to the OpenTelemetry internal
+ * diagnostic API
+ */
+export class DiagAPI implements DiagLogger {
+  /** Get the singleton instance of the DiagAPI API */
+  public static inst(): DiagAPI {
+    let theInst = null;
+    if (_global[GLOBAL_DIAG_LOGGER_API_KEY]) {
+      // Looks like a previous instance was set, so try and fetch it
+      theInst = _global[GLOBAL_DIAG_LOGGER_API_KEY]?.(
+        API_BACKWARDS_COMPATIBILITY_VERSION
+      ) as DiagAPI;
+    }
+
+    if (!theInst) {
+      theInst = new DiagAPI();
+      _global[GLOBAL_DIAG_LOGGER_API_KEY] = makeGetter(
+        API_BACKWARDS_COMPATIBILITY_VERSION,
+        theInst,
+        noopDiagApi()
+      );
+    }
+
+    return theInst;
+  }
+
+  /** Private internal constructor
+   * @private */
+  private constructor() {
+    let _logLevel: DiagLogLevel = DiagLogLevel.INFO;
+    let _filteredLogger: DiagLogger | null;
+    let _logger: DiagLogger = noopDiagLogger();
+
+    function _logProxy(funcName: keyof DiagLogger): DiagLogFunction {
+      return function () {
+        const orgArguments = arguments as unknown;
+        const theLogger = _filteredLogger || _logger;
+        const theFunc = theLogger[funcName];
+        if (theFunc && typeof theFunc === 'function') {
+          return theFunc.apply(
+            theLogger,
+            orgArguments as Parameters<DiagLogFunction>
+          );
+        }
+      };
+    }
+
+    // Using self local variable for minification purposes as 'this' cannot be minified
+    const self = this;
+
+    // DiagAPI specific functions
+
+    self.getLogger = (): DiagLogger => {
+      // Return itself if no existing logger is defined (defaults effectively to a Noop)
+      return _logger;
+    };
+
+    self.setLogger = (logger: DiagLogger): DiagLogger => {
+      const prevLogger = _logger;
+      if (prevLogger !== logger && logger !== self) {
+        // Simple special case to avoid any possible infinite recursion on the logging functions
+        _logger = logger || noopDiagLogger();
+        _filteredLogger = diagLogLevelFilter(_logLevel, _logger);
+      }
+
+      return prevLogger;
+    };
+
+    self.setLogLevel = (maxLogLevel: DiagLogLevel) => {
+      if (maxLogLevel !== _logLevel) {
+        _logLevel = maxLogLevel;
+        if (_logger) {
+          _filteredLogger = diagLogLevelFilter(maxLogLevel, _logger);
+        }
+      }
+    };
+
+    // DiagLogger implementation
+    const theFuncs: Array<keyof DiagLogger> = [
```

Comment:
Using an embedded array to avoid a for..in loop that could lead to incorrect functionality if the end-user is using some library that decorates objects. e.g. like ember.js and prototype.js